### ISSUE TITLE
ChromeTraceBuildListener leaking memory?

### ISSUE THREAD

**Issue by msridhar** [facebook/buck#1839]
We did a Buck build with many targets that starting running out of memory, and a heap dump showed `ChromeTraceBuildListener` holding on to more than 1.5GB of memory.  Looking closer, it seemed that its [`outputExecutor`](https://github.com/facebook/buck/blob/master/src/com/facebook/buck/event/listener/ChromeTraceBuildListener.java#L926) had accumulated a huge number of tasks in its queue that were holding on to memory.  Running the build with `--config log.chrome_trace_generation=false` fixed the issue.  It seems there may be a leak here that should be fixed.

**Comment by kageiit** [comment]
Compiling a single large `android_library` target generates a trace file of 4.07 GB!

This is when all the dependent rules are already cached. Just a single javac invocation on a single build target creating a huge log file is definitely suspicious

**Comment by kageiit** [comment]
cc @styurin @ttsugriy This seems pretty serious

**Comment by sbalabanov** [comment]
@msridhar I imagine this might happen on slow drives when trace listener's background thread cannot write events fast enough and starts accumulating jobs.
@kageiit It says it did 1099 jobs, which means 1099 build rule. Each build rule has steps, everything is recorded. You can however open trace file in any editor and check which events are prevailing in your case.

**Comment by kageiit** [comment]
@sbalabnov it also says 1 updated. The rest are cache hits. The other steps did not run

**Comment by msridhar** [comment]
@sbalabanov not sure what you consider a slow drive, but this was on an SSD on a MacBook Pro.  Also only three build tasks were running (I have a quad core), so there should have been a CPU thread available to do the writes.  I guess it's possible that IO is the bottleneck, but I suspect something else is going on.

**Comment by sbalabanov** [comment]
kageiit Build rules are still recorded I guess, at least one record per each :) I agree though that 4Gb is way more than I would expect. I do not see your trace, so please investigate what record types are really busting it.

**Comment by sbalabanov** [comment]
@msridhar The way how this listener works is not quite straightforward but still simple. Any message dispatched from message bus is converted into ChromeRecord, which only has primitive types, so once dispatched, no memory is held for complex types like build rules. ChromeRecord then is passed as a parameter to a single thread build executor's task which will
1) serialize it into json
2) write it into a file
for each record synchronously.

So, if you see that executor keeps piling up tasks it means it does not have enough bandwidth to complete 1) and 2) (either can potentially be a bottleneck). Because tasks are piling up, memory is wasted to keep tasks' context.

I do not have a repro so I can't investigate myself what the real issue is, but you can add some simple instrumentation to find it out. Depending on the issue the possible solution might be
- Reduce write rate, for example do not record steps
- Batch writes to disk
- Serialize with multiple threads, may be speed it up somehow

As I mentioned above, I'd try to debug / heap dump first to detect which record types create the most of the burden.

**Comment by msridhar** [comment]
Hi @sbalabanov, here is some more info from the heap dump I have.

* The `outputExecutor` is a `java.util.concurrent.ThreadPoolExecutor`, whose `workQueue` field is a `java.util.concurrent.LinkedBlockingQueue` with a `count` field of `16160`.
* The `item` of each node in the `LinkedBlockingQueue` is a `FutureTask`, whose `callable` is a lambda allocated [here](https://github.com/facebook/buck/blob/master/src/com/facebook/buck/event/listener/ChromeTraceBuildListener.java#L927).  The lambda captures a `ChromeTraceEvent`.
* Usually the `ChromeTraceEvent`s retain very little memory, but some of them retain ~1MB of memory.  In such cases, the `category` field is `"javac"`, the `name` field is `"analyze"`, and the memory is retained in the `args` map.  The entries in the `args` map are either (1) a key like `file 1021` with a value being a path to a source file, or (2) a key like `type 1021` with a value being a class name.  So all the memory seems to be in `String`s.

Hope that is helpful.  Let me know what other info I can provide.

**Comment by sbalabanov** [comment]
@msridhar the thing I mostly interested in, is to group all pending tasks by `category` and `name` and see counts. This is to find out the abuser. It can be done by parsing a heap dump for the collection you mentioned above, or may be by just looking to the resulting trace file, which should be plain json.
Clearly it's just too many events to process at the given rate.

**Comment by kageiit** [comment]
Finally got to the bottom of this. The trace file is inundated with javac phase events (parse/enter/analyze) etc. This is causing the exponential blow up of the trace file size. We have 2.5k java files in a single `android_library` target and compiling it alone (all dependencies available in cache) causes a 4.07 Gb trace file.

I commented out method bodies of

https://github.com/facebook/buck/blob/master/src/com/facebook/buck/jvm/java/JavacEventSinkToBuckEventBusBridge.java#L97

and 

https://github.com/facebook/buck/blob/master/src/com/facebook/buck/jvm/java/JavacEventSinkToBuckEventBusBridge.java#L110

Ran the build again, and the size of the trace file is 4.7 Mb. Seems like javac phase event reporting is causing a 3 orders of magnitude size increase in the amount of logs written to disk. I think it would be best to disable this behavior by default and let users configure turning it on if needed.

I tried this with both the errorprone javac and the vanilla jdk 8 javac with same results, so the choice of compiler does not seem to matter.

cc @jkeljo @styurin @ttsugriy
